File: application/src/main/java/com/troblecodings/launcher/javafx/ViewManager.java
package com.troblecodings.launcher.javafx;

import com.troblecodings.launcher.LauncherView;
import com.troblecodings.launcher.javafx.views.HomeView;
import com.troblecodings.launcher.javafx.views.SettingsView;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.scene.Node;
import javafx.scene.control.Label;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;

public class ViewManager {
    private final ObjectProperty<LauncherView> currentView = new SimpleObjectProperty<>(LauncherView.NONE);
    private final HashMap<LauncherView, Node> viewCache = new HashMap<>();
    private final Logger logger = LoggerFactory.getLogger(ViewManager.class);

    public LauncherView getCurrentView() {
        return currentView.get();
    }

    public void setCurrentView(LauncherView view) {
        if (view == currentView.get()) {
            logger.debug("Already on view {}", view);
            return;
        }

        logger.debug("Switching views: {} -> {}", currentView.get(), view);
        currentView.set(view);
    }

    public ReadOnlyObjectProperty<LauncherView> currentViewProperty() {
        return currentView;
    }

    public Node getViewNode( LauncherView view) {
        if(viewCache.containsKey(view)) {
            return viewCache.get(view);
        }

        viewCache.put(view, createView(view));
        return viewCache.get(view);
    }

    private Node createView( LauncherView view) {
        return switch (view) {
            case LauncherView.HOME -> new HomeView().build();
            case LauncherView.SETTINGS -> new SettingsView().build();
            default -> {
                var label = new Label("Scene not found!");
                label.setStyle("-fx-text-fill: crimson");
                yield label;
            }
        };
    }
}

File: src/main/java/com/troblecodings/launcher/javafx/ViewManager.kt
package com.troblecodings.launcher.javafx

import com.troblecodings.launcher.LauncherView
import com.troblecodings.launcher.javafx.views.HomeView
import com.troblecodings.launcher.javafx.views.LoginView
import com.troblecodings.launcher.javafx.views.SettingsView
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.scene.Node
import javafx.scene.control.Label
import org.slf4j.LoggerFactory

class ViewManager {
    private val currentView: ObjectProperty<LauncherView> = SimpleObjectProperty(LauncherView.NONE)
    private val viewCache: HashMap<LauncherView, Node> = HashMap()
    private val logger = LoggerFactory.getLogger(ViewManager::class.java)

    fun getCurrentView(): LauncherView {
        return currentView.get()
    }

    fun setCurrentView(view: LauncherView) {
        if (view == currentView.get()) {
            logger.debug("Already on view {}", view)
            return;
        }

        logger.debug("Switching views: {} -> {}", currentView.get(), view)
        currentView.set(view)
    }

    fun currentViewProperty(): ObjectProperty<LauncherView> {
        return currentView
    }

    fun getViewNode(view: LauncherView): Node {
        return viewCache.getOrPut(view) { createView(view) }
    }

    private fun createView(view: LauncherView): Node {
        return when (view) {
            LauncherView.HOME -> HomeView().build()
            LauncherView.SETTINGS -> SettingsView().build()
            else -> Label("Scene not found!").apply { style = "-fx-text-fill: crimson" }
        }
    }
}